Fix 03172_error_log_table_not_empty

pamarcos opened the pull request:
Fix 03172_error_log_table_not_empty without any time constraints
    
Let's get rid of the time constraints by checking the previous number of errors and making sure they're increased.

### Changelog category (leave one):
- CI Fix or Improvement (changelog entry is not required)

#65601

pamarcos committed:
Increase the interval to be less optimistic

Turns out when executing lots of tests in parallel
there might be more than 1 min elapsed between the
first errors and the second ones.

robot-ch-test-poll4 commented:
<!-- automatic status comment for PR #65604 from pamarcos/ClickHouse:03172_error_log_table_not_empty -->
*This is an automated comment for commit 059e7860b56e1287575d2ae82220fda69476b35f with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65604/059e7860b56e1287575d2ae82220fda69476b35f/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Cloud fork sync (only for ClickHouse Inc. employees)</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

devcrafter commented:
This PR makes the test less flaky but doesn't fix flakiness. The `system.errors` lacks columns which can provide context. I'd consider adding there `query_id` at least (it'll be empty if error occurs not in a query context). This way, `query_id` could be used to filter the errors, and it'd not be necessary to rely on time.

devcrafter reviewed (commented):
It can be a temporary fix. And I'd suggest addressing the comment above (can be in separate PR)

pamarcos commented:
>  The `system.errors` lacks columns which can provide context.

Just to clarify, do you mean `system.errors` or the new `system.error_log`?

>I'd consider adding there `query_id` at least (it'll be empty if error occurs not in a query context). This way, `query_id` could be used to filter the errors, and it'd not be necessary to rely on time.

I thought exactly the same, because it'd help to disambiguate the query without the need for filtering within any time window at all.

pamarcos commented:
I'm taking a look and we may have to limit the amount of query ids we store in `system.errors`. Maybe we can store the last 100 or 1000 or something among those lines.

e.g. the `NOT_IMPLEMENTED` in a recently started server in my dev machine has almost 20K entries
```sql
SELECT
    name,
    code,
    value
FROM system.errors

Query id: ff43780c-36c3-4081-bbfc-d0e1961a3a3c

   ┌─name─────────────────────┬─code─┬─value─┐
1. │ NOT_IMPLEMENTED          │   48 │ 19424 │
2. │ CANNOT_MANIPULATE_SIGSET │  111 │     1 │
3. │ NO_ELEMENTS_IN_CONFIG    │  139 │     2 │
4. │ KEEPER_EXCEPTION         │  999 │     2 │
   └──────────────────────────┴──────┴───────┘

4 rows in set. Elapsed: 0.016 sec. 
```

pamarcos committed:
Fix 03172_error_log_table_not_empty without any time constraints

Let's get rid of the time constraints by checking the previous
number of errors and making sure they're increased.

devcrafter commented:
@pamarcos Not a big deal here, but not sure that merging w/o review is a good thing. AFAIS, the updated test should be run sequentially, i.e. `# Tags: no-parallel` since it relies on previous state. Otherwise, in case of the test flakiness, it can succeed if another instance of the same test (running in parallel) wrote errors into log

pamarcos commented:
> # Tags: no-parallel

You're absolutely right. My apologies. I rushed it because I saw more instances of the flake test making harm to other PRs CI.
I've created [another PR ](https://github.com/ClickHouse/ClickHouse/pull/65725) adding the tag and added you as reviewer, if you don't mind. Please feel free to add any feedback there in case there's something missing apart from the `no-parallel` tag.

>And I'd suggest addressing the comment above (can be in separate PR)

Regarding this, I have the work done in [a branch](https://github.com/pamarcos/ClickHouse/tree/add-query-ids-to-system-errors) that adds last 10  `query_id`s to `system.errors` and `system.error_log`. However, [I presented the idea in a Slack thread](https://clickhouse-inc.slack.com/archives/C02F2LML5UG/p1719244050812499), but I didn't get any positive feedback about it.

CheSema commented:
That test `03172_error_log_table_not_empty` become flaky after that change.

CheSema commented on tests/queries/0_stateless/03172_error_log_table_not_empty.sh:
```diff
@@ -0,0 +1,41 @@
+SELECT throwIf(true, 'error_log', toInt16(111)) SETTINGS allow_custom_error_code_in_throwif=1; -- { serverError 111 }
+SELECT throwIf(true, 'error_log', toInt16(222)) SETTINGS allow_custom_error_code_in_throwif=1; -- { serverError 222 }
+SELECT throwIf(true, 'error_log', toInt16(333)) SETTINGS allow_custom_error_code_in_throwif=1; -- { serverError 333 }
+SELECT sleep(2) format NULL;
```
It should be marked as `Tags: no-fasttest` since it uses timeouts.

CheSema replied:
the second type of error is stale, it from the old version of the test ` AND event_time > now() - INTERVAL 1 MINUTE;`

CheSema replied:
Systems tables are created lazily. It could be not created yet when we read from it.

CheSema replied:
fixing it https://github.com/ClickHouse/ClickHouse/pull/66093

pamarcos replied:
>It should be marked as Tags: no-fasttest since it uses timeouts.

[Here's](https://github.com/ClickHouse/ClickHouse/pull/65725#discussion_r1654928104) the reasoning why I closed the PR that added precisely that tag

>the second type of error is stale, it from the old version of the test  AND event_time > now() - INTERVAL 1 MINUTE;

That was part of the original test. The timing issue was fixed precisely in this PR. Are you sure the error you see is not from a prior and flaky version of the test?

>Systems tables are created lazily. It could be not created yet when we read from it.

I don't follow you. At what point is the test reading from the table without flushing logs first?

CheSema replied:
> > It should be marked as Tags: no-fasttest since it uses timeouts.
> 
> [Here's](https://github.com/ClickHouse/ClickHouse/pull/65725#discussion_r1654928104) the reasoning why I closed the PR that added precisely that tag

No. You were trying to add `Tags: no-parallel`. I'm talking about `Tags: no-fasttest` because the test uses timeouts.

> > the second type of error is stale, it from the old version of the test  AND event_time > now() - INTERVAL 1 MINUTE;
> 
> That was part of the original test. The timing issue was fixed precisely in this PR. Are you sure the error you see is not from a prior and flaky version of the test?

That error from the test before you changes. Your changes fixed that error, that is good.

> > Systems tables are created lazily. It could be not created yet when we read from it.
> 
> I don't follow you. At what point is the test reading from the table without flushing logs first?

here
`errors_111=$($CLICKHOUSE_CLIENT -q "SELECT sum(value) FROM system.error_log WHERE code = 111")`

pamarcos replied:
>No. You were trying to add Tags: no-parallel. I'm talking about Tags: no-fasttest because the test uses timeouts.

Right, but since it's not using timeouts anymore it shouldn't be needed, right?

>here
errors_111=$($CLICKHOUSE_CLIENT -q "SELECT sum(value) FROM system.error_log WHERE code = 111")

You're absolutely right, my bad

CheSema replied:
> Right, but since it's not using timeouts anymore it shouldn't be needed, right?

Yes. This is why I fight for it and trying remove timeouts. It is better to leave the test to work as fast-test. Then mark it as `no-fasttest`.

pamarcos replied:
I see. The test takes ~6s to complete. If that's too long for a fast-test (I don't know where the limit is) let's add the `no-fasttest` :+1:
